### PR TITLE
check for patches not referenced in conandata.yml

### PR DESCRIPTION
**Pull request by ericLemanissier**
reimplement part of https://github.com/ericLemanissier/conan-center-lint-patches/blob/main/lint_patches.py
results are tracked in https://github.com/conan-io/conan-center-index/issues/21146

### CODE REVIEW

**Comment by ericLemanissier** [comment]
last PRs fixing offenders:
- https://github.com/conan-io/conan-center-index/pull/23739
- https://github.com/conan-io/conan-center-index/pull/23054
- https://github.com/conan-io/conan-center-index/pull/23155
- https://github.com/conan-io/conan-center-index/pull/23182
- https://github.com/conan-io/conan-center-index/pull/23353

**Comment by ericLemanissier** [comment]
@uilianries @RubenRBS Do you have a way to test this hook on all current references in CCI ? `conan cci:export-all-versions` is not available on conan 1, and this hook is not available in conan 2 :/

**Comment by uilianries** [comment]
Nice! I see it as a good improvement, but I would prefer fixing all recipes first, because it's classified as error. Otherwise, any non-related PR would be broken.

**Comment by ericLemanissier** [comment]
sure, I'll do all the PRs

File: hooks/conan-center.py [hunk 1]
```diff
@@ -943,6 +945,28 @@ def test(out):
...
+            for patch in patches:
+                if 'patch_file' not in patch:
+                    continue
+                patch_file_name = str(patch["patch_file"])
+                patch_file_name = os.path.relpath(patch_file_name)  # fixes the path (double slashes for example)
+                patch_file_name = patch_file_name[8:]
+                if patch_file_name in unused_patches:
+                    unused_patches.remove(patch_file_name)
+        for patch in unused_patches:
+            out.error(f"Following patch files is not referenced in conandata.yml: patches/{patch}")
```
Comment (Croydon) [inline hooks/conan-center.py:968]: Will this actually display several errors in CCI or only the first that exit the build? Maybe convert `unused_patches` from list to string and display all at once.

Also grammar: Either `file is` if you keep it one patch file at a time or `files are` if all at once.

Comment (ericLemanissier) [reply hooks/conan-center.py:968]: you can see the actual errors generated in my message above: they are grouped with one line for each file, eg
```
[HOOK - conan-center.py] pre_export(): ERROR: [NO DANGLING PATCHES (KB-H078)] Following patch files is not referenced in conandata.yml: patches/file1.patch (https://github.com/conan-io/conan-center-index/blob/master/docs/error_knowledge_base.md#KB-H078-NO-DANGLING-PATCHES) 
[HOOK - conan-center.py] pre_export(): ERROR: [NO DANGLING PATCHES (KB-H078)] Following patch files is not referenced in conandata.yml: patches/file2.patch (https://github.com/conan-io/conan-center-index/blob/master/docs/error_knowledge_base.md#KB-H078-NO-DANGLING-PATCHES) 
```
Thanks for the grammar correction !

Comment (Croydon) [reply hooks/conan-center.py]: Yes, it works in general, my question was regarding how CCI's CI is handling such hook errors and if they abort the build on the first one or if they can display multiple errors within a single hook before it gets abort.

Comment (ericLemanissier) [reply hooks/conan-center.py]: AFAIU the only way hooks can interrupt the flow is by raising exceptions. the only exceptions in https://github.com/conan-io/hooks/blob/HEAD/hooks/conan-center.py are raised at the end of a "step" (pre_export, post_export etc.) when [`raise_if_error_output`](https://github.com/conan-io/hooks/blob/0d81053e4b54eac6f6cb55a6f1358e62874b7857/hooks/conan-center.py#L157) wrapper calls [`raise_if_error`](https://github.com/conan-io/hooks/blob/0d81053e4b54eac6f6cb55a6f1358e62874b7857/hooks/conan-center.py#L154), so after each check of the step has finished, and only if one of the check reported a status with a level higher than `CONAN_HOOK_ERROR_LEVEL`

Comment (uilianries) [reply hooks/conan-center.py]: I agree with @Croydon , keeping it in a single a output would be much simpler.

Comment (ericLemanissier) [reply hooks/conan-center.py]: done

**Comment by ericLemanissier** [comment]
As you wanted, all the patches are now grouped on the same line instead of one line per patch.
Here are two examples of logs generated:
```
[HOOK - conan-center.py] pre_export(): ERROR: [NO DANGLING PATCHES (KB-H078)] Following patch file(s) are/is not referenced in conandata.yml: patches/2.13.0-0001-fix-fallthrough.patch (https://github.com/conan-io/conan-center-index/blob/master/docs/error_knowledge_base.md#KB-H078-NO-DANGLING-PATCHES)
```
```
[HOOK - conan-center.py] pre_export(): ERROR: [NO DANGLING PATCHES (KB-H078)] Following patch file(s) are/is not referenced in conandata.yml: patches/1.1.1.patch, patches/fix-cmake-source-dir-and-targets.patch, patches/pstring.patch, patches/strlen.patch (https://github.com/conan-io/conan-center-index/blob/master/docs/error_knowledge_base.md#KB-H078-NO-DANGLING-PATCHES)
```
I don't think it's more readable, but as you want

**Comment by ericLemanissier** [comment]
@uilianries all the fixes have been merged. The only offenders remaining are either deprecated, or not compatible with conan 2.

**Review by jcar87 — CHANGES_REQUESTED** [review]
Pending discussion with the team.
(likely to be merged, this is a good change, thanks @ericLemanissier )

**Comment by ericLemanissier** [comment]
@jcar87 What's the opinion of the team regarding this change ?

**Comment by ericLemanissier** [comment]
@jcar87 @uilianries can we please have an update on the status of this PR ?

**Comment by ericLemanissier** [comment]
https://github.com/conan-io/conan-center-index/pull/24165

**Pull request closed without merging**